### ISSUE TITLE
Struggling to Reproduce StackLLaMA

### ISSUE THREAD

**Issue by rajpabari** [huggingface/trl#641]
Hello, I am currently trying to reproduce the evaluation results in the [original StackLLaMA blog post](https://huggingface.co/blog/stackllama) after fine-tuning LLaMA with PPO, but am struggling to do so. My goal with reproducing this baseline is to ensure that I can accurately generate completions and evaluate the reward that a LLaMA model achieves on the Stack Exchange task.

Here are the steps I took for reproduction in order to reproduce the baseline from the blog post --

Compute available: 4x NVIDIA Tesla V100s (EC2 p3.8xlarge instance)

1. Make a fresh conda environment, in the conda environment pip install requirements.txt bitsandbytes datasets accelerate loralib
2. Download [huggyllama](https://huggingface.co/huggyllama/llama-7b/tree/main), [reward model](https://huggingface.co/kashif/llama-7b_stack-exchange_RM_peft-adapter-merged), and [official adapter weights after PPO-finetuning](https://huggingface.co/edbeeching/llama-se-rl-adapter).
3. Merge official LORA weights with huggyllama using provided merge_peft_adapter.py script
4. Run rl_training.py, with the following modifications to perform evaluation only (not training, given that we have loaded in the PPO-finetuned model):
- Change device_map to 'auto' everywhere rather than just being on current_device
- Comment out line 188, peft_config = lora_config -- we do not have a LORA head because we are only evaluating the model. After these changes, when we instantiate the model it looks like this:
```python
model = AutoModelForCausalLMWithValueHead.from_pretrained(
    config.model_name,
    load_in_8bit=True,
    device_map="auto", 
    #device_map={"": current_device},
    #peft_config=lora_config,
)
``` 
- Note, I also tried loading in the model using AutoModelForCausalLM (without the value head, given that we are not training anything) and achieved similar results
- In the training loop, introduce variables to calculate the running average reward, and remove the portion of the loop that steps the trainer. The modified main loop looks like this: 
```python
train_log = pd.DataFrame(columns=['texts','rewards'])
running_avg = 0
n = 0
with torch.no_grad():
    for epoch, batch in tqdm(enumerate(ppo_trainer.dataloader)):
        if epoch >= config.total_ppo_epochs:
            break

        question_tensors = batch["input_ids"]
        response_tensors = ppo_trainer.generate(
            question_tensors,
            return_prompt=False,
            length_sampler=output_length_sampler,
            **generation_kwargs,
        )
        n+=1
        batch["response"] = tokenizer.batch_decode(response_tensors, skip_special_tokens=True)

        # Compute reward score (using the sentiment analysis pipeline)
        texts = [q + r for q, r in zip(batch["query"], batch["response"])]
        pipe_outputs = sentiment_pipe(texts, **sent_kwargs)
        rewards = [torch.tensor(output[0]["score"] - script_args.reward_baseline) for output in pipe_outputs]

        #Introduce new logging statements
        train_log.loc[n-1] = [texts, rewards]
        running_avg += (rewards[0].item() - running_avg) / n
        print(n, running_avg)
        if n % 100 == 0:
            train_log.to_pickle('/home/ubuntu/stackllama_repro/trl/eval/train_log.pkl')
        # Run PPO step
        # stats = ppo_trainer.step(question_tensors, response_tensors, rewards)
        # ppo_trainer.log_stats(stats, batch, rewards)

        # if script_args.save_freq and epoch and epoch % script_args.save_freq == 0:
        #     ppo_trainer.save_pretrained(script_args.output_dir + f"step_{epoch}")
```
- Pass in batch_size of 2 in script_args, in order to not run into out-of-memory errors given compute constraints

I am running the script with the above modifications in an effort to reproduce the baseline from the blog post. However, the initial results are much worse than what I was expecting -- after 298 iterations, I've achieved an average reward of -0.574. From the [StackLLaMA blog post](https://huggingface.co/blog/stackllama), after training, we should be observing rewards no lower than -0.5 at the absolute minimum (see figure below).

![image](https://github.com/huggingface/trl/assets/52271154/77c277db-92b0-4a3c-a44e-f6172f014696)
 
I've also tried using the adapter weights that @mnoukhov pushed to the hub as a baseline, as linked in #401, as opposed to those from edbeeching and kashif that I linked above. However, I was consistently achieving a reward of NaN, so there must have been some issue incurred somewhere.

Given all of this, I tried using the rl_training.py script with even fewer modifications, to see if I could continue training the PPO-finetuned StackLLaMA. This time, the only changes I made were making the device_map to "auto" and making the batch_size 2, just as before (note, I did NOT comment out the LORA and training/stepping stuff this time). I logged the rewards at each iteration of the training loop and achieved similarly unexpected results as I did before.

Any insights/help on reproducing the results from the StackLLaMA blog post?

**Comment by rajpabari** [comment]
I unfortunately don't have access to an A100 at the moment

**Comment by mnoukhov** [comment]
The reward model is built on top of the supervised finetuned model. Are you merging weights correctly? see https://github.com/huggingface/trl/issues/401#issuecomment-1605578501

**Comment by rajpabari** [comment]
I got the reward model from [here](https://huggingface.co/kashif/llama-7b_stack-exchange_RM_peft-adapter-merged), which was already merged. I'll confirm this with your RM, I don't recall if, when I tried with the weights you pushed to the hub, I merged your RM into the base LLaMA or the supervised fine-tuned model.

**Comment by lvwerra** [comment]
Not sure that the reward model there from @kashif works. Can you merge the weights here: https://huggingface.co/trl-lib/llama-7b-se-rm-peft?

**Comment by kashif** [comment]
thanks @lvwerra having a look

**Comment by Andres-Carranza** [comment]
Hi @lvwerra  @mnoukhov, I'm a colalborator with @rajpabari on this and yes we did merge the reward models with the supervised fine-tuned model. We tried merging the weights published in #401 and also the weights from https://huggingface.co/trl-lib/llama-7b-se-rm-peft and both resulted in similarly negative results.

**Comment by BillChan226** [comment]
@rajpabari How do you calculate the running average reward?
